core/src/main/java/com/linecorp/armeria/server/AnnotatedHttpServiceMethod.java

```diff
@@ -142,44 +152,157 @@ private Object invoke(ServiceRequestContext ctx, HttpRequest req, @Nullable Aggr
+        ParamMap paramMap = null;
+        Object[] values = new Object[parameters.size()];
+        for (int i = 0; i < parameters.size(); ++i) {
+            Parameter entry = parameters.get(i);
+            final String value;
+            switch (entry.parameterType()) {
+                case PATH_PARAM:
+                    value = ctx.pathParam(entry.name());
+                    assert value != null;
+                    values[i] = convertParameter(value, entry.type());
+                    break;
+                case PARAM:
+                    if (paramMap == null) {
+                        paramMap = paramMapOf(ctx, req, message);
+                    }
+                    value = getParam(paramMap, entry);
+                    values[i] = value != null ? convertParameter(value, entry.type()) : null;
+                    break;
+                case PREDEFINED_TYPE:
+                    if (entry.type().isAssignableFrom(ServiceRequestContext.class)) {
+                        values[i] = ctx;
+                    } else if (entry.type().isAssignableFrom(HttpRequest.class)) {
+                        values[i] = req;
+                    } else if (entry.type().isAssignableFrom(AggregatedHttpMessage.class)) {
+                        values[i] = message;
+                    } else if (entry.type().isAssignableFrom(ParamMap.class)) {
+                        if (paramMap == null) {
+                            paramMap = paramMapOf(ctx, req, message);
+                        }
+                        values[i] = paramMap;
+                    }
+                    break;
             }
         }
-        return parameters;
+        return values;
     }
 
-    BiFunction<ServiceRequestContext, HttpRequest, Object> withConverter(ResponseConverter converter) {
-        return (ctx, req) ->
-                executeSyncOrAsync(ctx, req).thenApply(obj -> convertResponse(obj, converter));
+    /**
+     * Returns a map of parameters decoded from a request.
+     */
+    private static ParamMap paramMapOf(ServiceRequestContext ctx, HttpRequest req,
+                                       @Nullable AggregatedHttpMessage message) {
+        try {
+            if (req.method() == HttpMethod.GET) {
+                return wrap(new QueryStringDecoder(ctx.query(), false).parameters());
+            }
+
+            String contentType = req.headers().get(HttpHeaderNames.CONTENT_TYPE);
+            if (contentType == null || contentType.isEmpty()) {
+                return EMPTY_MAP;
+            }
+
+            MediaType mediaType = AggregationStrategy.aggregationAvailable(contentType);
+            if (mediaType == null) {
+                return EMPTY_MAP;
+            }
+
+            assert message != null;
+            String body = message.content().toStringAscii();
+            return wrap(new QueryStringDecoder(body, mediaType.charset().orElse(ISO_8859_1), false)
```

Comment:
It seems like it's encoded as UTF-8 when it appears in a URI:

- https://www.ietf.org/rfc/rfc3986.txt (SO: https://stackoverflow.com/a/5433216) 

For x-www-form-urlencoded, it seems slightly more complicated than URI:

- https://www.w3.org/TR/html5/forms.html#application/x-www-form-urlencoded-encoding-algorithm (See the decoding part)